### ISSUE TITLE
sclang: errors with make_parser.sh and Bison 3.0.4

### ISSUE THREAD

**Issue by nhthn** [supercollider/supercollider#2259]
```
$ bison -V
bison (GNU Bison) 3.0.4
Written by Robert Corbett and Richard Stallman.

Copyright (C) 2015 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
```

After running `lang/LangSource/Bison/make_parser.sh`, thus updating `lang11d_tab.cpp`, compilation succeeds. But when I run sclang, I get a huge number of errors like this:

```
ERROR: syntax error, unexpected BINOP, expecting NAME or WHILE
  in file '/home/nathan/.local/share/SuperCollider/Extensions/sc3-plugins-sc/LoopBufUGens/sc/classes/LJP Classes/Scale.sc'
  line 3 char 7:

        var <>stepRatios;
       ^^

-----------------------------------
ERROR: file '/home/nathan/.local/share/SuperCollider/Extensions/sc3-plugins-sc/LoopBufUGens/sc/classes/LJP Classes/Scale.sc' parse failed
error parsing
ERROR: syntax error, unexpected BINOP, expecting NAME or WHILE
  in file '/home/nathan/.local/share/SuperCollider/downloaded-quarks/wslib/wslib-classes/Extensions/Various/Main-monitorOSC.sc'
  line 8 char 12:

        classvar <>exclude;
            ^^

-----------------------------------
ERROR: file '/home/nathan/.local/share/SuperCollider/downloaded-quarks/wslib/wslib-classes/Extensions/Various/Main-monitorOSC.sc' parse failed
error parsing
ERROR: syntax error, unexpected BINOP, expecting NAME or WHILE
  in file '/home/nathan/git/supercollider/SCClassLibrary/JITLib/GUI/ProxyMonitorGui.sc'
  line 8 char 12:

        classvar <>lastOutBus;
            ^^
```

All of the errors seem to stem from class variable declarations of the form `var <>foo` and `classvar <>foo`.

**Comment by scztt** [comment]
I don't think this is necessarily a bug - @snappizz, you modified the lang11d before doing this, yes?

**Comment by snappizz** [comment]
Nope, I stashed those changes before trying `./make_parser.sh`.

**Comment by brianlheim** [comment]
Just tried this; can't recreate. Can someone else try again?

**Comment by brianlheim** [comment]
Was able to recreate this using Bison 3.0.4. It's OK using Bison 2.3.

**Comment by brianlheim** [comment]
Ok I think I figured this out! Issue is with make_parser.sh. My fix was as follows:

1. Add `-d` option to bison command, which creates a .hpp header file to accompany the .cpp:
```
```
2. Delete the .h file and change the include in PyrLexer.cpp to `lang11d_tab.hpp`

**Comment by brianlheim** [comment]
We should update the files, but only after #52 is solved or closed.

**Comment by snappizz** [comment]
Thanks, that seems to work. Is this change backward-compatible with Bison 2?

**Comment by brianlheim** [comment]
Great! I've tested with v2.3 and v3.0.4, and they both generate .hpp headers, so I think it's back-compatible.

**Comment by snappizz** [comment]
I think you should submit two PRs -- one for this issue, and one for #2259. You can either wait for the first one to merge before starting the second one, or file two PRs with one dependent on the other.

**Comment by brianlheim** [comment]
Ok, sure. Do you mind if I leave the current one up just for testing/discussion purposes?

**Comment by snappizz** [comment]
Yes, that's fine.

**Comment by eleses** [comment]
Frankly having a non-recompilable parser (from the Bison source) except with an ancient version of Bison, seems a rather nasty state of affairs. (After all, this project aims to be compatible with and using e.g. recent Boost or recent Qt versions.) Since the class-extensions-anywhere feature doesn't seem to be progressing, I think this bug should be fixed in preference to that, i.e. first, if at all possible.

**Comment by brianlheim** [comment]
it's not about compatibility with bison, that was a red herring. the script is just wrong, it needs to create both the header and cpp file -- https://github.com/supercollider/supercollider/issues/2259#issuecomment-271775511

**Comment by eleses** [comment]
Yeah, the script probably never worked because I don't see any mention in the Bison news that `%defines` (or `-d`) was changed to be required so that the header is generated... Speaking of which: why is that script even needed, since it's a trivial rule to put a Makefile (of whatever flavor)?

**Comment by brianlheim** [comment]
it could be integrated into the build system, that script predates moving to CMake

**Comment by eleses** [comment]
N.B. The tab.cpp file currently included in the tree was generated with Bison 2.4.1 (there's a comment in it to that effect.) I think Bison was called with the longer `--defines` so that a `.h` was generated instead of hpp one, although the default extension for C++ headers. may have changed with Bison versions.

**Comment by brianlheim** [comment]
i'm fine with any of these solutions so long as the command produces compilable code :)
